File: haskell/day01/Main.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE MultilineStrings #-}

module Main where

import Control.Monad (when)
import Text.Printf (printf)
import Control.Monad.State

testFile :: FilePath
testFile = "./day01/test.txt"

inputFile :: FilePath
inputFile = "./day01/input.txt"

type Step = Either Int Int
type Dial = State Int Int

parseStep :: String -> Step
parseStep = \case
    'L' : n -> Left  $ read n
    'R' : n -> Right $ read n
    _ -> error "invalid input"

step2int :: Step -> Int
step2int (Left  n) = -n
step2int (Right n) =  n

first :: (a -> a') -> (a, b) -> (a', b)
first f (a, b) = (f a, b)

second :: (b -> b') -> (a, b) -> (a, b')
second f (a, b) = (a, f b)

addMod100 :: Integral a => a -> a -> a
addMod100 a b = (a + b) `mod` 100

parseInput :: FilePath -> IO [Int]
parseInput fpath =
    map (step2int . parseStep)
    . filter (not . null)
    . lines
    <$> readFile fpath

answer1 :: [Int] -> Int
answer1 xs = length . filter (== 0) $ scanl addMod100 initial xs

count :: Step -> Dial
count step = do
    current <- get
    let step' = step2int step
    let new = current + step'
    let cond =
            if current == 0
                then new <= (-100) || new >= 100
                else new <= 0      || new >= 99
    put $ new `mod` 100
    let
        one :: Float
        one = fromIntegral new / 100
    if cond
        then return . abs $ floor one
        else return 0

count2 :: Step -> Dial
count2 (Right step) = do
    current <- get
    let new = current + step
    put $ new `mod` 100
    if (step + current) >= 100
        then return $ (step + current) `div` 100
        else return 0
count2 (Left step) = do
    current <- get
    let new = current - step
    put $ new `mod` 100
    let cond = if current == 0 then step >= 100 else step >= current
    if cond
        then return $ (step - current) `div` 100 + 1
        else return 0

count2' :: Step -> Dial
count2' step = do
    current <- get
    let
        delta = step2int step
        next = current + delta
        cross = if delta >= 0
                then next `div` 100 - current `div` 100
                else (current - 1) `div` 100 - (next - 1) `div` 100
    put next
    return cross

answer2 :: [Step] -> Int
answer2 input = sum $ evalState (mapM count2' input) initial

stepping :: Step -> Dial
stepping step = do
    current <- get
    let new = current + step2int step
    put new
    return . abs $
        new `div` 100 - current `div` 100

initial :: Int
initial = 50

main :: IO ()
main = do
    tests <- parseInput testFile
    when (answer1 tests /= 3) do
        error "algorithm is wrong in answer1"

    inputs <- parseInput inputFile
    let answer01 = answer1 inputs

    tests2 <- map parseStep . lines <$> readFile testFile
    when (answer2 tests2 /= 6) do
        error "algorithm is wrong in answer2"

    inputs2 <- map parseStep . filter (not . null) . lines <$> readFile inputFile
    let answer02 = answer2 inputs2

    printf """
        ====================
        DAY 01
        Answer 1 = %d
        Answer 2 = %d
        ====================
        """
        answer01
        answer02

File: rust/src/day01/main.rs
use std::{error::Error, fs, str::FromStr};

const INPUT_FILE: &str = "src/day01/input.txt";
const TEST_FILE: &str = "src/day01/test.txt";
const DIAL_MODULO: i64 = 100;

#[derive(Debug)]
enum Step {
    R(i64),
    L(i64),
}

impl Step {
    /// Mengembalikan nilai delta (positif atau negatif)
    fn as_delta(&self) -> i64 {
        match self {
            Self::R(n) => *n,
            Step::L(n) => -n,
        }
    }
}

impl FromStr for Step {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut chars = s.chars();
        let dir = chars.next().ok_or("Input kosong")?;
        let num_str = chars.as_str();

        let num: i64 = num_str
            .parse()
            .map_err(|_| format!("Bukan angka: {num_str}"))?;

        match dir {
            'R' => Ok(Self::R(num)),
            'L' => Ok(Self::L(num)),
            c => Err(format!("arah tidak valid: {c}")),
        }
    }
}

fn parse_input(input: &str) -> Result<Vec<Step>, Box<dyn Error>> {
    input
        .lines()
        .map(|l| l.parse::<Step>().map_err(|e| e.into()))
        .collect()
}

fn answer1(steps: &[Step], initial: i64) -> i64 {
    let mut current = initial;
    let mut counter = 0;

    for step in steps {
        current = (current + step.as_delta()).rem_euclid(DIAL_MODULO);

        if current == 0 {
            counter += 1;
        }
    }

    counter
}

/// gemini's
fn answer2(steps: &[Step], initial: i64) -> i64 {
    let mut current_abs: i64 = initial;
    let mut counter: i64 = 0;

    for step in steps {
        let delta = step.as_delta() as i64;
        let next_abs = current_abs + delta;

        let cross = if delta >= 0 {
            next_abs.div_euclid(DIAL_MODULO) - current_abs.div_euclid(DIAL_MODULO)
        } else {
            (current_abs - 1).div_euclid(DIAL_MODULO) - (next_abs - 1).div_euclid(DIAL_MODULO)
        };

        counter += cross;
        current_abs = next_abs;
    }

    counter
}

// fn answer2(steps: &[Step], initial: i32) -> i32 {
//     let mut dial = Dial::new(initial);
//     let mut counter = 0;
//
//     for step in steps {
//         let overflow = dial.add_overflow(step);
//         match overflow {
//             Some(o) => counter += o,
//             None => {
//                 if dial.arrow == 0 {
//                     counter += 1;
//                 }
//             },
//         }
//
//     }
//
//     counter
// }

fn main() -> Result<(), Box<dyn Error>> {
    let tests_str = fs::read_to_string(TEST_FILE)?;
    let inputs_str = fs::read_to_string(INPUT_FILE)?;

    let tests = parse_input(&tests_str)?;
    let inputs = parse_input(&inputs_str)?;

    let test = answer1(&tests, 50);
    assert_eq!(test, 3, "algorithm is wrong in answer 1");

    let answer01 = answer1(&inputs, 50);

    let test = answer2(&tests, 50);
    assert_eq!(test, 6, "algorithm is wrong in answer 2");

    let answer02 = answer2(&inputs, 50);

    println!(
        "==================== \n\
        Day 01                \n\
        Answer 1 = {answer01} \n\
        Answer 2 = {answer02} \n\
        ===================="
    );

    Ok(())
}
